### ISSUE TITLE
Repeated measures for species: conceptual problems with the random effects

### ISSUE THREAD

**Issue by ErikRingen** [ScottClaessens/coevolve#84]
I was thinking more about the use of specific random effect parameters for repeated measures at the species level, and I am not sure it makes sense given the rest of the model. Because each tip species already gets its own parameter eta with phylogenetically independent drift. So just thinking from a generative modelling perspective, does it really make sense that species change along the terminal branch, and then some extra species-specific variance is added on top of that? I don't think ignoring repeated measures is problematic at the species level because each species already has its own parameter with unique drift terms. The dynamical model is not really amenable to the same type of variance decomposition as GLMMs.

In contrast, if there were repeated measures of the same individual within species, or different subgroups/populations measured within a species, we would want to have random effects for that nested structure.

**Comment by ScottClaessens** [comment]
Hmm, so are you basically saying that even without the additional random effect, the model implies some species-specific variance anyway, due to the terminal drift?

I guess I'm viewing it more like a standard phylogenetic multilevel model, where the extra random effect just captures all the species-specific variance that is _not_ due to phylogenetic history. Like what Paul says in his brms vignette on the topic [here](https://cran.r-project.org/web/packages/brms/vignettes/brms_phylogenetics.html#a-phylogenetic-model-with-repeated-measurements).

**Comment by ScottClaessens** [comment]
I've only tested a few. The model from the vignette seems to fit reasonably well:

```r
coev_fit(
    data = repeated$data,
    variables = list(
      x = "normal",
      y = "normal"
    ),
    id = "species",
    tree = repeated$phylogeny,
    # additional arguments for cmdstanr
    parallel_chains = 4,
    iter_warmup = 2000,
    iter_sampling = 2000,
    refresh = 0,
    seed = 1
  )
```

**Comment by ErikRingen** [comment]
Thanks for sharing the example. It converges fine but the results are telling of the problem. Imagine first we just took the species means:

```
repeated_mean <- repeated$data %>% 
  group_by(species) %>% 
  summarise(x = mean(x), y = mean(y))

fit_mean <- coev_fit(
  data = repeated_mean,
  variables = list(
    x = "normal",
    y = "normal"
  ),
  id = "species",
  tree = repeated$phylogeny,
  # additional arguments for cmdstanr
  parallel_chains = 4,
  iter_warmup = 250,
  iter_sampling = 250,
  seed = 1
)
```

In this model, there is a strong drift correlation and a cross-selection effect from x -> y. However, if you fit the example above, the species random effect correlation consumes both of those effects, driving them to zero. In contrast, this does *not* happen if you fit a typical PGLMM in brms:

```
library(brms)
A <- ape::vcv.phylo(repeated$phylogeny, corr = T)

model_repeat1 <- brm(
  bf(mvbind(scale(x), scale(y)) ~ (1|sp_rep) + (1|p|gr(species, cov = A))),
  data = repeated$data %>% mutate(sp_rep = species),
  family = gaussian(),
  data2 = list(A = A),
  chains = 4, cores = 4,
  iter = 500, warmup = 250,
  backend = "cmdstanr"
)
```

**Comment by ScottClaessens** [comment]
Huh, okay I see! So, what do we think the solution is? We could just remove the random effect.

**Comment by ErikRingen** [comment]
I still like the functionality. Perhaps we could allow the grouping variables to be added as named variables rather than automatically detected? And then generate random intercepts for each?

**Comment by ScottClaessens** [comment]
Hmm, with arbitrary group effects, we start getting into the kind of territory where it would be preferable to have actual formulas for each response variable. Like brms, we would add random effects when `(1 | group)` and could even add arbitrary predictor variables. But this starts to get complicated quite quickly, and would require a change in the way we have set things up.

But yes, as you say, it could just be a vector/list of named categorical variables in the dataset:

```r
coev_fit(
  data = d,
  variables = list(
    x = "bernoulli_logit",
    y = "normal"
  ),
  id = "id",
  tree = tree,
  random = c("group1", "group2")
)
```

**Comment by ErikRingen** [comment]
Yeah I dont think we should accommodate the more complex formulas at this time. The syntax up propose above looks good, perhaps we disable the repeated measures for now until we can implement the new system.

**Comment by ScottClaessens** [comment]
Thinking about this a bit more today. Currently, the way the model works is that it applies an independent drift term to every _observation_ in the dataset (i.e., it iterates over rows in the model block). So that means that if we have a dataset with repeated observations for species, each observation of that species will have a different drift term added to the eta parameter. With this model setup, I agree that random effects for each species do not make a lot of sense.

But do we want the model to be set up this way? Would it not be more realistic to give each _species_, rather than each _observation_, its own independent drift term? In other words, do we want to assume that the within-species variation we see is purely due to drift on the phylogenetic tree, or due to other non-phylogenetic differences between individuals? I feel like the latter makes more sense, and it's the logic that underlies repeated measures analyses in PGLMMs.

In other words, I feel like we're interested in modelling the dynamic coevolution of _species means_, with any observation-level variation around those means due to non-phylogenetic reasons.

But I could be thinking about this all wrong. What do you think @ErikRingen?

**Comment by ErikRingen** [comment]
Hi Scott,

Thanks for coming back around to this. 

Yeah, we definitely should not have independent drift terms for each observation, that doesn't make sense. You're on point about the evolution of species means. In a PGLMM, y = phylogenetic covariance + species-specific variance + observation-specific variance. Without repeated observations, we cannot partition species-specific variance from obs-level variance. This is why the Gaussian requires special handling in the pointwise likelihood, because there is no disentanglement of the species-specific and observation-level variances.

Thinking about our model, it is clear that everything leading up to the terminal branch is the analog of phylogenetic covariance. The changes along the terminal branch (both selection and drift) should constitute the species-specific variances. For the observation-specific variance, I think it will depend on the distribution of the response. For non-Gaussian variables,  I don't think any changes are necessary--the latent etas for each species just become better identified by the repeated observations, and the observation-level variance is determined by the dispersion/shape/scale parameters. For Gaussian traits, I think it would be: y ~ Normal(eta_tip, residual_sigma). So there is an additional parameter to estimate.

To recap, analogizing to PGLMMs, we are already estimating the phylogenetic covariance and species-specific variances (which are modelled via random effects in PGLMMs). For Gaussian traits with no repeat measures, the latter is confounded by the observation-level variance.

**Comment by ScottClaessens** [comment]
Yes, this makes sense!

For the point about independent drift terms for each species rather than each observation, do the attached changes to the model start to move us in the right direction? This is the model that the package produces for two Gaussian traits with repeated measures. You can use https://www.diffchecker.com/ to see where I've made changes in the new version.

I managed to fit this model to the `repeated` dataset mentioned above. The model fits well, though the cross-selection effects are still estimated to be zero, likely because the model still includes species-level random effects.

[new_stancode_repeated_gaussian.txt](https://github.com/user-attachments/files/18049146/new_stancode_repeated_gaussian.txt)
[old_stancode_repeated_gaussian.txt](https://github.com/user-attachments/files/18049147/old_stancode_repeated_gaussian.txt)

**Comment by ErikRingen** [comment]
Hi Scott,

OK so now the sigma_group represents within-group variance, as opposed to the normal meaning in which sigma_group would refer to variance of the group means. We should definitely document this carefully so avoid confusion. 

Looking at your code, the covariance matrix for the likelihood at the tips will no longer be VCV_tips. Instead, it should be the "group" level covariance matrix, and the correlated terminal drift we can use the standard non-centered paramaterization that we do for all the internal nodes and models with strictly non-Gaussian responses. So in the case of repeated measures we can once again treat the terminal drift as latent variables, if that makes sense.

**Comment by ScottClaessens** [comment]
I think I understand. Lots of moving parts here. Is it worth writing out the model for both cases? My attempt:

#### Two Gaussian traits - no repeated observations

$$
\begin{align*}
\text{residualx}\_i = x_i - \eta_{xi} \\\\
\text{residualy}\_i = y_i - \eta_{yi} \\\\
\begin{pmatrix} \text{residualx}_i \\\\ \text{residualy}_i \end{pmatrix} \sim \text{MVNormal}(0, \text{VCVtips}\_{tip[i]})
\end{align*}
$$

#### Two Gaussian traits - repeated observations

$$
\begin{align*}
\text{residualx}\_i = x_i - (\eta_{xi} + \text{tdriftx}\_i) \\\\
\text{residualy}\_i = y_i - (\eta_{yi} + \text{tdrifty}\_i) \\\\
\begin{pmatrix} \text{tdriftx}_i \\\\ \text{tdrifty}_i \end{pmatrix} \sim \text{MVNormal}(0, \text{VCVtips}\_{tip[i]}) \\\\
\begin{pmatrix} \text{residualx}_i \\\\ \text{residualy}_i \end{pmatrix} \sim \text{MVNormal}(0, \Sigma)
\end{align*}
$$

**Comment by ErikRingen** [comment]
Yeah, that looks correct. For non-Gaussian traits I think we also end up modelling observation-level random effects, because we need to capture the observation-level correlations between traits (not just the species-level correlations).

**Comment by ScottClaessens** [comment]
Okay, cool. This is enough for me to go on, I think! So rather than removing the repeated measures functionality, as we had discussed, I'll make some changes in this direction. We can discuss as we go.

It would be good to get this sorted before we submit, so I'll work on this over the coming week.

**Comment by ScottClaessens** [comment]
Hey, I had a first go at this today. I removed the missing data ifelse statements to reduce some of the complexity in the model block. Here's the model:

[new_stancode_repeated_gaussian.txt](https://github.com/user-attachments/files/18062246/new_stancode_repeated_gaussian.txt)

Unfortunately, the model fails to converge, with high Rhats for the `terminal_drift` and `Q_sigma` parameters. It's possible that the model is struggling to distinguish between variation due to drift on the tree and species-specific residual variation. But it's also likely I've made a mistake somewhere. Could you take a look @ErikRingen? Sorry - I'm not an expert when it comes to writing Stan code!

If this continues to be an issue, perhaps we could for now just go ahead with the simpler model where we have one terminal drift term per tip but no random effects.

```r
# code to test model
library(coevolve)
tree <- ape::rcoal(10)
d <- data.frame(
  id = rep(tree$tip.label, each = 5),
  x = rnorm(50),
  y = rnorm(50)
)
data_list <- coev_make_standata(
  data = d,
  variables = list(
    x = "normal",
    y = "normal"
  ),
  id = "id",
  tree = tree
)
library(rstan)
model <- rstan::stan_model(file = "new_stancode_repeated_gaussian.stan")
fit <- sampling(model, data = data_list)
```

**Comment by ErikRingen** [comment]
Thanks! And the refactor on missing data looks good.

Fortunately, the code had an easy fix: it was missing a prior on the terminal_drift parameters. Attaching the updated code.

[new_stancode_repeated_gaussian_fixed.txt](https://github.com/user-attachments/files/18073051/new_stancode_repeated_gaussian_fixed.txt)

```
library(coevolve)

rpt_standat <- coev_make_standata(
  data = repeated$data,
  variables = list(
    x = "normal",
    y = "normal"
  ),
  id = "species",
  tree = repeated$phylogeny
)

new_mod <- cmdstanr::cmdstan_model("new_stancode_repeated_gaussian.stan")

fit_new <- new_mod$sample(
  data = rpt_standat,
  chains = 6,
  parallel_chains = 6,
  iter_warmup = 200,
  iter_sampling = 200,
  adapt_delta = 0.98,
  refresh = 10
)

fit_new$summary("Q")
fit_new$summary("cor_R")
fit_new$summary("cor_group")
```

Now that we have a working version, I think we need to decide:

- What to call these parameters. As I mentioned, sigma_group is likely to throw people off, because owing to the structure of the model this must refer to *within* species variation. So perhaps "sigma_species" and then careful explanation?

- Whether this should be implemented automatically or not. I am inclined to think that there should be an argument (true by default) that gives users the option *not* to estimate these parameters if they don't want to. This would give them the flexibiliy to fit a simpler model, e.g., for model comparison. It might also be useful when the repeated measures are very sparse, imagine a case where only a few species have just a couple repeats, then it may be very hard to identify these parameters.

**Comment by ErikRingen** [comment]
> Thanks! And the refactor on missing data looks good.
> 
> Fortunately, the code had an easy fix: it was missing a prior on the terminal_drift parameters. Attaching the updated code.
> 
> [new_stancode_repeated_gaussian_fixed.txt](https://github.com/user-attachments/files/18073051/new_stancode_repeated_gaussian_fixed.txt)
> 
> ```
> library(coevolve)
> 
> rpt_standat <- coev_make_standata(
>   data = repeated$data,
>   variables = list(
>     x = "normal",
>     y = "normal"
>   ),
>   id = "species",
>   tree = repeated$phylogeny
> )
> 
> new_mod <- cmdstanr::cmdstan_model("new_stancode_repeated_gaussian.stan")
> 
> fit_new <- new_mod$sample(
>   data = rpt_standat,
>   chains = 6,
>   parallel_chains = 6,
>   iter_warmup = 200,
>   iter_sampling = 200,
>   adapt_delta = 0.98,
>   refresh = 10
> )
> 
> fit_new$summary("Q")
> fit_new$summary("cor_R")
> fit_new$summary("cor_group")
> ```
> 
> Now that we have a working version, I think we need to decide:
> 
> * What to call these parameters. As I mentioned, sigma_group is likely to throw people off, because owing to the structure of the model this must refer to _within_ species variation. So perhaps "sigma_species" and then careful explanation?
> * Whether this should be implemented automatically or not. I am inclined to think that there should be an argument (true by default) that gives users the option _not_ to estimate these parameters if they don't want to. This would give them the flexibiliy to fit a simpler model, e.g., for model comparison. It might also be useful when the repeated measures are very sparse, imagine a case where only a few species have just a couple repeats, then it may be very hard to identify these parameters.

Or perhaps "sigma_residual" or "sigma_within" is even clearer naming.

**Comment by ScottClaessens** [comment]
Thanks for this @ErikRingen!

Just to clarify, I removed the missing data ifelse statements just to get this example model running. We will want to keep that functionality in the package version.

Out of interest, when you fitted this model to the `repeated` dataset, did you get the same answer as you did when you fitted the model to the species means only? It would be good at some point to fit this repeated observations model to a dataset where we know the right answer, just to check it's working properly.

Regarding your points, I like the name "sigma_residual". We can then treat these parameters as residual variance and correlations and refer to them in a similar way to brms (e.g., rescor; see multivariate brms output below).

```
 Family: MV(gaussian, gaussian) 
  Links: mu = identity; sigma = identity
         mu = identity; sigma = identity 
Formula: x ~ 1 
         y ~ 1 
   Data: d (Number of observations: 20) 
  Draws: 4 chains, each with iter = 2000; warmup = 1000; thin = 1;
         total post-warmup draws = 4000

Regression Coefficients:
            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
x_Intercept    -0.13      0.18    -0.49     0.23 1.00     4390     2509
y_Intercept    -0.03      0.27    -0.56     0.51 1.00     3773     2577

Further Distributional Parameters:
        Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
sigma_x     0.80      0.15     0.58     1.14 1.00     4440     2619
sigma_y     1.21      0.21     0.88     1.69 1.00     4132     2824

Residual Correlations: 
            Estimate Est.Error l-95% CI u-95% CI Rhat Bulk_ESS Tail_ESS
rescor(x,y)     0.12      0.22    -0.33     0.54 1.00     3286     2611

Draws were sampled using sampling(NUTS). For each parameter, Bulk_ESS
and Tail_ESS are effective sample size measures, and Rhat is the potential
scale reduction factor on split chains (at convergence, Rhat = 1).
```

Then, yes, I agree we should have an argument in `coev_fit()` that is set to TRUE by default. How about this?

```r
coev_fit(
  data = d,
  variables = list(
    x = "normal",
    y = "normal"
  ),
  id = "id",
  tree = tree,
  estimate_residual = TRUE
)
```

We can produce a warning similar to when missing data are detected, telling users that they can turn off this behaviour by setting `estimate_residual = FALSE`.

**Comment by ErikRingen** [comment]
Yes, it produces very similar results to the species-averaged model (but of course we do not expect it to be identical). Yeah `estimate_residual` sounds fine to me, with the caveat that it is only utilized when there are some repeated observations per species. So would `NULL` be a better default value then?

**Comment by ScottClaessens** [comment]
Hmm, yeah I see what you mean. I think TRUE is fine though, as long as we make it clear that the argument only applies when there are repeated observations. This is also the case for `complete_cases = TRUE` with/without missing data.

**Comment by ScottClaessens** [comment]
Okay, I'll start implementing this. And then in the case of non-Gaussian variables, we would prefer to use the non-centered parameterisation, right?

**Comment by ErikRingen** [comment]
Yes. And with a mixture we need the centered param, as before.

**Comment by ScottClaessens** [comment]
Okay, I've had a go at making these changes this weekend. But I'm struggling with the distinction between the model with one or more Gaussian variables and the model without any Gaussian variables (e.g., the Bernoulli-Bernoulli model).

You say that in the former case, the `sigma_residual` parameter refers to the _within_-species variance. But in the latter case, we use the non-centered parameterisation and so the model becomes something like:

```stan
for (i in 1:N_obs) {
  vector[N_tree] lp = rep_vector(0.0, N_tree);
  for (t in 1:N_tree) {
    vector[J] tdrift;
    vector[J] residuals;
    tdrift = cholesky_decompose(VCV_tips[t, tip_id[i]]) * to_vector(terminal_drift[t][tip_id[i],]);
    residuals = cholesky_decompose(diag_pre_multiply(sigma_residual, L_residual)) * to_vector(residual_z[i,]);
    if (miss[i,1] == 0) lp[t] += bernoulli_logit_lpmf(to_int(y[i,1]) | eta[t,tip_id[i]][1] + tdrift[1] + residuals[1]);
    if (miss[i,2] == 0) lp[t] += bernoulli_logit_lpmf(to_int(y[i,2]) | eta[t,tip_id[i]][2] + tdrift[2] + residuals[2]);
  }
  target += log_sum_exp(lp);
}
```

But isn't this just the random effects model that we had before? So `sigma_residual` now refers again to the "variance of the group means"?

I wonder if we're not overcomplicating things somewhat by trying to model the residual variation. Perhaps a better way to proceed is to remove the random effects (as we initially discussed) but still allow repeated observations for species. Having multiple observations will just help the model to better estimate the species-level `eta` and `terminal_drift` parameters. What do you think @ErikRingen?

**Comment by ErikRingen** [comment]
Yeah in the case of non-Gaussian variables, it would capture observational-level covariance. The unique variance should be captured by the dispersion terms in the case of Gamma, Beta etc. An intermediate case is like the Poisson, where it would also do double duty of capturing over-dispersion.

I would be in favor of keeping them, if it not too much of a headache to code up.

**Comment by ScottClaessens** [comment]
Okay, so they are basically correlated ORLEs. This makes sense, thank you!

Yes, I'll give it another go. Just lots of ifelse statements in `coev_make_stancode()`, but should be doable.

**Comment by ScottClaessens** [comment]
Right, here's my first attempt. I've removed the generated quantities block for now, but I will add it again later.

Here are the models that `coev_make_stancode()` produces for different mixtures of variables with repeated observations:

[repeated_bernoulli_bernoulli.txt](https://github.com/user-attachments/files/18151539/repeated_bernoulli_bernoulli.txt)
[repeated_gaussian_gaussian.txt](https://github.com/user-attachments/files/18151554/repeated_gaussian_gaussian.txt)
[repeated_gaussian_bernoulli.txt](https://github.com/user-attachments/files/18151557/repeated_gaussian_bernoulli.txt)

I've fitted the first two models and they seem to run fine. I haven't yet run the third model, as I'm not sure that the `residuals[2] = tdrift[2]` code is correct. Could you take a look @ErikRingen?

Another thing to consider is that the repeated observation models with Gaussian variables do not currently handle missing data at all. I wasn't sure how to generalise the previous missing data code to the repeated case. It would be good to find a general solution that will work for both non-repeated and repeated data.

**Comment by ScottClaessens** [comment]
Thanks! I'll implement these changes and make sure the third model fits well.

Regarding missing data, we could use code like we had before, so something like this:

```stan
if (miss[i,1] == 0) {
  tdrift[1] = y[i,1] - eta[t,tip_id[i]][1];
  terminal_drift[t][tip_id[i],1] ~ std_normal();
} else {
  tdrift[1] = terminal_drift[t][tip_id[i],1];
}
```

We set it up this way to avoid having `terminal_drift` parameters without priors. This setup makes sense if there are same number of tips and observations. But in the case of repeated observations, wouldn't this set the prior for `terminal_drift[t][tip_id[i],1]` multiple times? And then what if we have cases where there is missing data for only some observations within a particular species? It's possible that I'm overthinking this though.

**Comment by ScottClaessens** [comment]
At least according to this blog post, declaring multiple priors for the same parameter shouldn't be an issue: https://discourse.mc-stan.org/t/multiple-priors-for-the-same-parameter/10943/9

**Comment by ErikRingen** [comment]
> Thanks! I'll implement these changes and make sure the third model fits well.
> 
> Regarding missing data, we could use code like we had before, so something like this:
> 
> ```stan
> if (miss[i,1] == 0) {
>   tdrift[1] = y[i,1] - eta[t,tip_id[i]][1];
>   terminal_drift[t][tip_id[i],1] ~ std_normal();
> } else {
>   tdrift[1] = terminal_drift[t][tip_id[i],1];
> }
> ```
> 
> We set it up this way to avoid having `terminal_drift` parameters without priors. This setup makes sense if there are the same number of tips and observations. But in the case of repeated observations, wouldn't this set the prior for `terminal_drift[t][tip_id[i],1]` multiple times? And then what if we have cases where there is missing data for only some observations within a particular species? It's possible that I'm overthinking this though.

But in the case of repeated measures, tdrift should not be y[i,1] - eta[t,tip_id[i]][1]? It should be like in the gaussian-gaussian model code you sent:

```
tdrift = cholesky_decompose(VCV_tips[t, tip_id[i]]) * to_vector(terminal_drift[t][tip_id[i],]);
residuals[1] = y[i,1] - (eta[t,tip_id[i]][1] + tdrift[1]);
```

**Comment by ScottClaessens** [comment]
Sorry, I'm confusing things. This is actually for the model with no repeated observations. I have separated the tdrift and residuals terms to make their distinction clearer. Here are the (current) model blocks for Gaussian traits with and without repeated observations:

### Without repeated observations

```stan
model{
  b ~ std_normal();
  for (t in 1:N_tree) {
    eta_anc[t] ~ std_normal();
    for (i in 1:(N_seg - 1)) z_drift[t, i] ~ std_normal();
  }
  A_offdiag ~ std_normal();
  A_diag ~ std_normal();
  L_R ~ lkj_corr_cholesky(4);
  Q_sigma ~ std_normal();
  if (!prior_only) {
    for (i in 1:N_obs) {
      vector[N_tree] lp = rep_vector(0.0, N_tree);
      for (t in 1:N_tree) {
        vector[J] tdrift;
        if (miss[i,1] == 0) {
          tdrift[1] = y[i,1] - (eta[t,tip_id[i]][1]);
          terminal_drift[t][tip_id[i],1] ~ std_normal();
        } else {
          tdrift[1] = terminal_drift[t][tip_id[i],1];
        }
        if (miss[i,2] == 0) {
          tdrift[2] = y[i,2] - (eta[t,tip_id[i]][2]);
          terminal_drift[t][tip_id[i],2] ~ std_normal();
        } else {
          tdrift[2] = terminal_drift[t][tip_id[i],2];
        }
        lp[t] = multi_normal_cholesky_lpdf(tdrift | rep_vector(0.0, J), cholesky_decompose(VCV_tips[t, tip_id[i]]));
      }
      target += log_sum_exp(lp);
    }
  }
}
```

### With repeated observations

```stan
model{
  b ~ std_normal();
  for (t in 1:N_tree) {
    eta_anc[t] ~ std_normal();
    for (i in 1:(N_seg - 1)) z_drift[t, i] ~ std_normal();
    to_vector(terminal_drift[t]) ~ std_normal();
  }
  A_offdiag ~ std_normal();
  A_diag ~ std_normal();
  L_R ~ lkj_corr_cholesky(4);
  Q_sigma ~ std_normal();
  // priors for residual sds and cors
  sigma_residual ~ exponential(1);
  L_residual ~ lkj_corr_cholesky(2);
  if (!prior_only) {
    for (i in 1:N_obs) {
      vector[N_tree] lp = rep_vector(0.0, N_tree);
      for (t in 1:N_tree) {
        vector[J] tdrift;
        vector[J] residuals;
        tdrift = cholesky_decompose(VCV_tips[t, tip_id[i]]) * to_vector(terminal_drift[t][tip_id[i],]);
        residuals[1] = y[i,1] - (eta[t,tip_id[i]][1] + tdrift[1]);
        residuals[2] = y[i,2] - (eta[t,tip_id[i]][2] + tdrift[2]);
        lp[t] = multi_normal_cholesky_lpdf(residuals | rep_vector(0.0, J), diag_pre_multiply(sigma_residual, L_residual));
      }
      target += log_sum_exp(lp);
    }
  }
}
```

Currently, the latter model does not deal with missing data. But I'm now thinking that actually the missing data functionality can just look the same as above.

**Comment by ScottClaessens** [comment]
Hi @ErikRingen, working on this again today. I'm still just focusing on the model block to make sure everything is working correctly. Here are the models again with missing data functionality and the changes you suggested:

[repeated2_bernoulli_bernoulli.txt](https://github.com/user-attachments/files/18164758/repeated2_bernoulli_bernoulli.txt)
[repeated2_gaussian_gaussian.txt](https://github.com/user-attachments/files/18164760/repeated2_gaussian_gaussian.txt)
[repeated2_gaussian_bernoulli.txt](https://github.com/user-attachments/files/18164764/repeated2_gaussian_bernoulli.txt)

However, the third model does not seem to fit very well. Do you have any idea why?

**Comment by ErikRingen** [comment]
Not sure yet. It wouldn't explain the poor fit in this case because I assume you're testing on a single tree, but the prior for residual_z shouldn't be inside the tree loop. Because there is only one set of shared residual_z parameters for all trees, applying the prior density N_tree times will result in a very different (much more concentrated) prior.

**Comment by ScottClaessens** [comment]
Yes, I tested on a single tree.

Okay, this makes sense. Let me know how you think this should be rewritten.

**Comment by ErikRingen** [comment]
We could just move it outside the tree loop, where the rest of the prior declarations are:

```
  for (i in 1:N_obs) {
    if (miss[i,1] == 0) residual_z[1,i] ~ std_normal();
  }
```

I haven't solved this problem, but some progress is that I have identified sigma_residual[2] as the problem. See example:

```
library(coevolve)
library(dplyr)
library(cmdstanr)

d_rpt <- coevolve::repeated
d_rpt$data$y <- as.integer(ifelse(d_rpt$data$y > median(d_rpt$data$y), 1, 0))

standat <- coevolve::coev_make_standata(data = d_rpt$data, variables = list(x = "normal", y = "bernoulli_logit"), id = "species", tree = d_rpt$phylogeny )

rpt_gaussian_bern <- cmdstan_model("repeated2_gaussian_bernoulli.stan")

rpt_fit <- rpt_gaussian_bern$sample(
  data = standat,
  chains = 4,
  parallel_chains = 4,
  iter_sampling = 300,
  iter_warmup = 300
)

summary_fit <- rpt_fit$summary()
summary_fit %>% arrange(desc(rhat))

draws_sigma <- posterior::as_draws_df(rpt_fit$draws(c("sigma_residual", "Q_sigma")))
bayesplot::mcmc_trace(rpt_fit$draws(c("sigma_residual", "Q_sigma", "L_residual")))
```

**Comment by ScottClaessens** [comment]
Yep, that prior declaration works.

Yeah, `sigma_residual[2]` and also all of `residual_z[2,i]` seem to not be behaving well.

**Comment by ErikRingen** [comment]
Pragmatically, maybe for now we print a warning to users if they try repeated with mix of gaussian/non-gaussian that it is experimental? Then I can circle back when I have more bandwidth to figure this one out.

**Comment by ScottClaessens** [comment]
Okay, sounds good. I'll push forward with this and let you know if anything else comes up.

**Comment by ScottClaessens** [comment]
Hi @ErikRingen, just checking that this is how the repeated gaussian-gaussian model should look with measurement error? I used the same approach as before.

[repeated_me_gaussian_gaussian.txt](https://github.com/user-attachments/files/18197704/repeated_me_gaussian_gaussian.txt)

**Comment by ErikRingen** [comment]
I don't think that is right, because the standard deviations aren't additive and the additivity doesn't hold when working with the cholesky decomposition:

```
R <- matrix(c(1, 0.5, 0.5, 1), 2, 2, byrow=T) # correlation mat
lower_R <- t(chol(R)) # cholesky_corr
sigma_res <- c(3, 4) # residual sds
sigma_ME <- c(0.2, 0.15) # measurement error SEs

# This is what the covariance matrix should be (additive variances)
cov_plus_me <- diag(sigma_ME^2) + (diag(sigma_res) %*% R %*% diag(sigma_res))
cov_plus_me
chol_cov_plus_me <- t(chol(cov_plus_me))

# test whether this is the same thing as Scott's code
chol_scott_code <- diag(sigma_ME) + (diag(sigma_ME) %*% lower_R)

# compare
chol_cov_plus_me
chol_scott_code
```

I would construct the residual covariance matrix like this:

```
matrix[J, J] residual_cov = diag_matrix(square(sigma_se)) + quad_form_diag(L_residual * L_residual', sigma_residual)
```

Then:

```
lp[t] = multi_normal_cholesky_lpdf(residuals | rep_vector(0.0, J), cholesky_decompose(residual_cov))
```

**Comment by ScottClaessens** [comment]
Thanks! I changed to your suggestion and the model fits well. Does the previous approach still work for `VCV_tips` in models without repeated observations?

Slowly getting there... now I'm just working on the generated quantities block.

**Comment by ErikRingen** [comment]
This looks right. And it is a good question about the measurement error in yrep. I think both including it and not would be defensible, depending on what you wanted to do with the model checks. Perhaps we ignore it for now.

**Comment by ScottClaessens** [comment]
Phew, okay, I think we're there. Here are the Stan models with no repeated observations and with repeated observations for different mixtures of variables. When we set `estimate_residual = FALSE`, the package uses the former model even when repeated observations are detected.

I have set `log_lik = TRUE` for all of these so you can see how the log-likelihood calculations are done in the generated quantities.

[norepeated3_bernoulli_bernoulli.txt](https://github.com/user-attachments/files/18211314/norepeated3_bernoulli_bernoulli.txt)
[norepeated3_gaussian_bernoulli.txt](https://github.com/user-attachments/files/18211315/norepeated3_gaussian_bernoulli.txt)
[norepeated3_gaussian_gaussian.txt](https://github.com/user-attachments/files/18211316/norepeated3_gaussian_gaussian.txt)
[repeated3_bernoulli_bernoulli.txt](https://github.com/user-attachments/files/18211318/repeated3_bernoulli_bernoulli.txt)
[repeated3_gaussian_bernoulli.txt](https://github.com/user-attachments/files/18211319/repeated3_gaussian_bernoulli.txt)
[repeated3_gaussian_gaussian.txt](https://github.com/user-attachments/files/18211320/repeated3_gaussian_gaussian.txt)

I have tested all of these and they fit without errors. @ErikRingen could you let me know what you think? If you're happy, I can write the tests and make sure the `summary()` output is correct. Then we should be good to go with this.

**Comment by ScottClaessens** [comment]
By the way, I tested the repeated observations model to make sure that it gets the right answer. It does.

```r
library(coevolve)
library(rstan)
# simulate coevolution where x -> y but not vice versa
sim <- coev_simulate_coevolution(
  n = 100,
  variables = c("x", "y"),
  selection_matrix = matrix(c(0.95, 0, 0.95, 0.95), byrow = TRUE, nrow = 2,
                            dimnames = list(c("x", "y"), c("x", "y"))),
  drift = c("x" = 0.05, "y" = 0.05),
  prob_split = 0.05
)
# repeated observations with some small uncorrelated error
d <- data.frame(
  id = rep(sim$data$species, each = 3),
  x = rep(sim$data$x, each = 3) + rnorm(300, sd = 0.1),
  y = rep(sim$data$y, each = 3) + rnorm(300, sd = 0.1)
)
# fit model
variables <- list(x = "normal", y = "normal")
scode <- coev_make_stancode(d, variables, "id", sim$tree)
sdata <- coev_make_standata(d, variables, "id", sim$tree)
model <- rstan::stan_model(model_code = scode)
fit <- rstan::sampling(model, sdata, cores = 4)
post <- rstan::extract(fit)
hist(post$A[,2,1]) # x -> y, positive
hist(post$A[,1,2]) # y -> x, zero
```

**Comment by ErikRingen** [comment]
I am happy with these, thanks for all the work to make this happen!
